### PR TITLE
Examples of solving two gym envs with AZ

### PR DESCRIPTION
**Pull request by davebiagioni**

### CODE REVIEW

File: examples/gym/solve_hallway.py [hunk 1]
```diff
@@ -19,60 +21,67 @@
...
-    def __init__(self, **kwargs):
-        super().__init__(gym.envs.make("CartPole-v0"), **kwargs)
-    
+    def __init__(self, size=16, max_steps=16):
+        env = HallwayEnv(size=size, max_steps=max_steps)
+        super().__init__(env)
+
     def get_obs(self) -> np.ndarray:
-        return np.array(self.state)
+        return self.env.get_obs()
```
Comment (pstjohn) [inline examples/gym/solve_hallway.py:33]: `AlphaZeroGymEnv` seems to want `get_env_obs` as the abstract method?

File: examples/gym/gym_state.py [hunk 2]
```diff
@@ -1,61 +1,57 @@
...
 import gym
 
 from rlmolecule.tree_search.graph_search_state import GraphSearchState
 
 from alphazero_gym import AlphaZeroGymEnv
 
+logger = logging.getLogger(__name__)
+
 
 class GymEnvState(GraphSearchState):
```
Comment (pstjohn) [inline examples/gym/gym_state.py:18]: would we want to add these classes to the main rlmolecule package?

File: examples/gym/hallway_env.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+if __name__ == "__main__":
+    env = HallwayEnv(16, 16)
+    done = False
+    _ = env.reset()
+    while not done:
+        action = env.action_space.sample()
+        print("action:", action)
+        obs, rew, done, _ = env.step(action)
+        print("result:", obs, rew, done)
```
Comment (pstjohn) [inline examples/gym/hallway_env.py:54]: I'm certainly happy to make a package directory that includes some of these testing environments (in fact I think hallway is already in tests/), this would be a good unittest though.

**Review by pstjohn — COMMENTED** [review]
This looks good! I definitely think some of the re-usable gym stuff could live in rlmolecule/gym or similar. 

just thinking about a trajectory towards an eventual readthedocs page, could the hallway example live (and run from) a jupyter notebook? it might be nice to have a single, self-contained example like that.

This commit history is pretty crazy, although I wasn't easily able to `rebase` of the current devel branch. I suppose the squash and merge should flatten a lot of these overlapping edits though?

File: examples/gym/solve_hallway.py (hunk 1 above)
Comment (davebiagioni) [reply examples/gym/solve_hallway.py:33]: `get_env_obs` is out of date, renamed to `get_obs`

File: examples/gym/gym_state.py (hunk 2 above)
Comment (davebiagioni) [reply examples/gym/gym_state.py:18]: yep, done!

File: examples/gym/hallway_env.py (hunk 3 above)
Comment (davebiagioni) [reply examples/gym/hallway_env.py:54]: i haven't written / pushed any unittests yet but the example run scripts are working with the latest changes

File: examples/gym/gridworld_env.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import logging
+from typing import Tuple
+
+import numpy as np
+import gym
+
+
+ACTION_MAP = {0: (0, 1), 1: (1, 0), 2: (0, -1), 3: (-1, 0)}
+OBSTACLE_CHANNEL = 0
```
Comment (pstjohn) [inline examples/gym/gridworld_env.py:9]: is there an obstacle in this implementation? might be a cool way to demo masked actions

File: examples/gym/solve_gridworld.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        return self.env.get_obs()
+
+
+class GridWorldProblem(GymProblem, TFAlphaZeroProblem):
+    def __init__(self, *,
+                 env: gw.GridEnv,
+                 engine: "sqlalchemy.engine.Engine",
+                 reward_class: Reward,
+                 **kwargs) -> None:
+        super().__init__(reward_class=reward_class, engine=engine, env=env, **kwargs)
```
Comment (pstjohn) [inline examples/gym/solve_gridworld.py:43]: i don't think this `__init__` definition actually doing anything, since it just passes the kwargs to the lower classes? I think this should run if you don't define an `__init__` here.

File: examples/gym/solve_gridworld.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                              hidden_dim=64)
+
+    def get_policy_inputs(self, state: GymEnvState) -> dict:
+        return {
+            "obs": state.env.get_obs(),
+            "steps": 0.*np.array([np.float64(self.env.episode_steps)])
+        }
+
+    def get_reward(self, state: GymEnvState) -> Tuple[float, dict]:
+        return state.cumulative_reward, {}
```
Comment (pstjohn) [inline examples/gym/solve_gridworld.py:62]: Interesting... any reason this couldn't get moved up to GymProblem?

File: examples/gym/solve_gridworld.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        for i in range(args.num_workers - 2):
+            jobs += [multiprocessing.Process(target=run_games)]
+
+        jobs += [multiprocessing.Process(target=train_model)]
+
+        for job in jobs[1:]:
+            job.start()
+
+        for job in jobs:
+            job.join(300)
```
Comment (pstjohn) [inline examples/gym/solve_gridworld.py:200]: todo for me, unless some else wanted to try it: does joblib's [Parallel](https://joblib.readthedocs.io/en/latest/parallel.html) (with the loky backend) throw the same tensorflow error if we don't delay imports to construct_problem()? That might simplify some of the problem initialization

File: examples/gym/tf_model.py [hunk 8]
```diff
@@ -21,3 +25,61 @@ def policy_model(obs_dim: int,
...
+        x = layers.Dense(hidden_dim, activation=activation)(x)
+    
+    value_logit = layers.Dense(1, name="value")(x)
+    pi_logit = layers.Dense(1, name="prior")(x)
+
+    return tf.keras.Model([obs], [value_logit, pi_logit], name="policy_model_cnn")
+
+
+
+def policy_model_2(obs_dim: Tuple[int],
```
Comment (pstjohn) [inline examples/gym/tf_model.py:58]: any reason for the two models? was the first one just for testing to see if `steps` is needed as an input?

File: rlmolecule/gym/gym_state.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    def __repr__(self) -> str:
+        return (self.env.get_obs(), self.step_count).__repr__()
+
+    def equals(self, other: any) -> bool:
+        are_close = np.all(np.isclose(self.env.get_obs(), other.env.get_obs()))
+        same_time = (self.step_count == other.step_count)
+        return are_close and same_time
+
+    def hash(self) -> int:
+        return hash(self.__repr__())
```
Comment (pstjohn) [inline rlmolecule/gym/gym_state.py:45]: `__repr__` of a numpy array probably looses a lot of information that you might need -- it only prints the first & last few rows and columns.
I'm assuming Charles wrote [this hash](https://github.com/NREL/rlmolecule/blob/1732e57acc1e0c990ad79b875524dd30573e2dea/rlmolecule/molecule/molecule_state.py#L54)

I did some quick [googling](https://madhatter106.github.io/DataScienceCorner/posts/speeding-up-numpy-array-comparison/), this might work?
`hash(self.env.get_obs().to_bytes()) ^ (13 * self.step_count)`

**Review by pstjohn — COMMENTED** [review]
This looks great! Happy to merge this as is, I left a few comments where we could probably fine-tune this a bit though

File: examples/gym/solve_gridworld.py (hunk 6 above)
Comment (davebiagioni) [reply examples/gym/solve_gridworld.py:62]: another good catch.  if we stick to the definition of env.step as always returning the step / terminal reward (which is what env.step is supposed to do) then yes ... i didn't initially migrate the hallway env properly to fit this definition.  but done now!

File: examples/gym/solve_gridworld.py (hunk 7 above)
Comment (davebiagioni) [reply examples/gym/solve_gridworld.py:192]: i can take a swing at this tomorrow, either way is fine with me. i was also going to add a notebook version of the gridworld problem so we have an easily runnable, interactive, self contained example

File: examples/gym/tf_model.py (hunk 8 above)
Comment (davebiagioni) [reply examples/gym/tf_model.py:58]: oh, actually, this might still be unresolved.  we have this issue around the state representation non-uniqueness leading to cycles.  the question is how to address this -- in the hallway example, the steps were already part of the state obs.  but in gridworld the obs is just the array, so i added steps here to enforce DAG structure.  i think in general the get_obs method implemented by a user could lead to cycles

File: rlmolecule/gym/gym_state.py (hunk 9 above)
Comment (davebiagioni) [reply rlmolecule/gym/gym_state.py:45]: ok, yeah i'm with you on the np array repr, hashing the bytestring makes sense.  but what is this bitwise xor stuff doing?

File: examples/gym/gridworld_env.py (hunk 4 above)
Comment (davebiagioni) [reply examples/gym/gridworld_env.py:9]: hah, no obstacles yet, i just left a few incomplete hooks -- right now the step method doesn't take this channel into account.  but it would be easy to enable this, and yeah that would be a great way to demo masked actions.

File: rlmolecule/gym/gym_state.py (hunk 9 above)
Comment (pstjohn) [reply rlmolecule/gym/gym_state.py:45]: no idea, i actually mis-read that as an exponentiation 😦 . @CharlesEdisonTripp, did you add that to the molecule state?
We just need a way to add that info into the hash. Maybe `hash((self.env.get_obs().to_bytes(), self.step_count))`?

Comment (davebiagioni) [reply rlmolecule/gym/gym_state.py:45]: yeah, i think that should work!  charles i am also curious about your xor voodoo : )

Comment (CharlesEdisonTripp) [reply rlmolecule/gym/gym_state.py]: Yeah was just an easy way to add in the _forced_terminal value to the hash of the string repr. A simpler and probably just as effective way would have been to add 13 * _forced_terminal (because _forced_terminal is boolean python will treat it as a 1 or 0 here). The xor is mostly just fancy in this case.. although I wanted to avoid the possibility of overflowing to beyond 64 bits.

A more general way to combine hash functions that usually works pretty well is to add them, multiplying each hash by some prime number, for example:  hash(a) + 13 * hash(b) + 31 * hash(c). In python's case of big integers, you might round it out with a bitwise and to clip it down to 64 bits: (hash(a) + 13 * hash(b) + 31 * hash(c)) & (2 ** 64 - 1).

Comment (CharlesEdisonTripp) [reply rlmolecule/gym/gym_state.py]: You could also avoid the clipping problem with XOR's, likely while retaining good hash properties:
hash(a) ^ (13 * hash(b)) ^ (31 * hash(c)). Extra points if you use a big "good hash table" prime (for example, see https://planetmath.org/goodhashtableprimes):

hash(a) ^ (805306457 * hash(b)) ^ (402653189 * hash(c))

If the hash functions are really orthogonal, you could safely xor them together without the primes:

hash(a) ^ hash(b) ^ hash(c)

But the issue with doing this typically is that if these were (say) all strings with the same hash function, simply xoring them together will give the same hash value for different permutations of the three strings. That's why we multiply by a prime. We use a prime because it typically gives better dispersion over the table than an ordinary number, particularly with power-of-two sized hash tables, which are typical for performance, memory management, and implementation reasons.

Comment (pstjohn) [reply rlmolecule/gym/gym_state.py]: Just to follow up, python does some binary XOR under the hood when you hash a tuple
